### DISCUSSION TITLE
v0.3.0

### DISCUSSION THREAD

**Discussion by andydunstall** [andydunstall/piko#71] [Announcements]
v0.3.0 contains a rewrite of the agent to server protocol. Instead of forwarding individual HTTP requests and responses, the protocol transports raw TCP bytes multiplexed over a single outbound connection from the upstream to the server.

This has a few major benefits:
- It supports non-HTTP protocols in the future (particularly TCP)
- It supports WebSockets
- It supports a Go SDK in the future where clients can listen for connections from Piko like any other TCP listener (such as `net.Listener` in Go)
- It makes the agent and server much simpler as they can treat the 'connection' to the upstream as any other TCP connection (even though the connection is actually multiplexed over the upstreams outbound connection to Piko)

This release also adds TLS support, both for the server and for the agent to configure custom root CAs to verify the servers TLS certificates. It also refactors the `piko agent` command line and configuration.

## What's Changed
* docs: add logo by @andydunstall in https://github.com/andydunstall/piko/pull/46
* client: add go sdk interface by @andydunstall in https://github.com/andydunstall/piko/pull/47
* agentv2: add agent cli and configuration by @andydunstall in https://github.com/andydunstall/piko/pull/48
* agentv2: add endpoint interface by @andydunstall in https://github.com/andydunstall/piko/pull/49
* agentv2: add endpoint reverse proxy by @andydunstall in https://github.com/andydunstall/piko/pull/50
* conn: add websocket options by @andydunstall in https://github.com/andydunstall/piko/pull/52
* tls: add tls support by @andydunstall in https://github.com/andydunstall/piko/pull/53
* agentv2: add heartbeats by @andydunstall in https://github.com/andydunstall/piko/pull/54
* docs: update logo by @andydunstall in https://github.com/andydunstall/piko/pull/55
* client: register endpoint listener by @andydunstall in https://github.com/andydunstall/piko/pull/56
* serverv2: add reverse proxy to forward requests to upstream by @andydunstall in https://github.com/andydunstall/piko/pull/57
* agentv2: handle listener connections by @andydunstall in https://github.com/andydunstall/piko/pull/58
* add docker package by @andydunstall in https://github.com/andydunstall/piko/pull/59
* tests: use httptest.Server by @andydunstall in https://github.com/andydunstall/piko/pull/60
* serverv2: add reverse proxy by @andydunstall in https://github.com/andydunstall/piko/pull/61
* agentv2: add pkg/mux by @andydunstall in https://github.com/andydunstall/piko/pull/62
* agentv2: add connection per listener by @andydunstall in https://github.com/andydunstall/piko/pull/63
* agentv2: add config options by @andydunstall in https://github.com/andydunstall/piko/pull/64
* serverv2: rewrite reverse proxy by @andydunstall in https://github.com/andydunstall/piko/pull/65
* serverv2: load balance among connected upstreams by @andydunstall in https://github.com/andydunstall/piko/pull/66
* piko: integrate agent/server v2 by @andydunstall in https://github.com/andydunstall/piko/pull/67
* admin: add request forwarding by @andydunstall in https://github.com/andydunstall/piko/pull/68
* status: add upstream endpoints status command by @andydunstall in https://github.com/andydunstall/piko/pull/69
* server: add usage reporting by @andydunstall in https://github.com/andydunstall/piko/pull/70

**Full Changelog**: https://github.com/andydunstall/piko/compare/v0.2.0...v0.3.0

<hr /><em>This discussion was created from the release <a href='https://github.com/andydunstall/piko/releases/tag/v0.3.0'>v0.3.0</a>.</em>

**Comment by randalee** [comment]
I upgraded the Helm-based deployment on EKS from version 0.2 to 0.3. I simply changed the provided image tag, but the pod failed to start.

```shell
{"level":"info","ts":"2024-06-12T01:29:39.709Z","subsystem":"main","msg":"starting piko server","node-id":"piko-0-lrwuftl","version":"v0.3.0"}
{"level":"info","ts":"2024-06-12T01:29:39.709Z","subsystem":"gossip","msg":"starting gossip","node-id":"piko-0-lrwuftl","bind-addr":":8003","advertise-addr":"172.20.6.189:8003"}
{"level":"warn","ts":"2024-06-12T01:29:39.716Z","subsystem":"main","msg":"failed to join cluster","error":"resolve: piko:8003: lookup host: piko: lookup piko on 10.100.0.10:53: no such host"}
{"level":"info","ts":"2024-06-12T01:29:39.717Z","subsystem":"proxy","msg":"starting proxy server","addr":"[::]:8000"}
{"level":"info","ts":"2024-06-12T01:29:39.717Z","subsystem":"admin","msg":"starting admin server","addr":"[::]:8002"}
{"level":"info","ts":"2024-06-12T01:29:39.717Z","subsystem":"admin","msg":"starting upstream server","addr":"[::]:8001"}
{"level":"warn","ts":"2024-06-12T01:29:40.761Z","subsystem":"gossip","msg":"failed to join cluster","error":"resolve: piko:8003: lookup host: piko: lookup piko on 10.100.0.10:53: no such host"}
{"level":"warn","ts":"2024-06-12T01:29:43.031Z","subsystem":"gossip","msg":"failed to join cluster","error":"resolve: piko:8003: lookup host: piko: lookup piko on 10.100.0.10:53: no such host"}
```


In the case of version 0.2, there are no issues on EKS, and for version 0.3, there are no problems when running it locally with Docker.
```bash
 docker run ghcr.io/andydunstall/piko:v0.3.0 server
{"level":"info","ts":"2024-06-12T01:50:33.555Z","subsystem":"main","msg":"starting piko server","node-id":"pbkwevb","version":"v0.3.0"}
{"level":"info","ts":"2024-06-12T01:50:33.555Z","subsystem":"gossip","msg":"starting gossip","node-id":"pbkwevb","bind-addr":":8003","advertise-addr":"10.10.0.2:8003"}
{"level":"info","ts":"2024-06-12T01:50:33.556Z","subsystem":"admin","msg":"starting upstream server","addr":"[::]:8001"}
{"level":"info","ts":"2024-06-12T01:50:33.556Z","subsystem":"admin","msg":"starting admin server","addr":"[::]:8002"}
{"level":"info","ts":"2024-06-12T01:50:33.556Z","subsystem":"proxy","msg":"starting proxy server","addr":"[::]:8000"}
```

**Reply by andydunstall** [reply]
Hi @randalee, sorry I removed the `/ready` route by mistake, so if the pod never becomes ready it won't be added to the service so DNS never resolves

I've merged a short term fix, and will have a look at improving how joining a cluster works in K8S next

**Reply by andydunstall** [reply]
I've released v0.4.0 with the fix so helm should now work, you'll still see `failed to join cluster` logs in the first pod as it takes a few seconds for the DNS record to be created by K8S, but it should eventually succeed

**Reply by randalee** [reply]
> I've released v0.4.0 with the fix so helm should now work, you'll still see `failed to join cluster` logs in the first pod as it takes a few seconds for the DNS record to be created by K8S, but it should eventually succeed

Thank you for your prompt feedback. Version 0.4 works well on EKS! 💯
